build/Extensions/DepsJsonExtensions.cs

```diff
@@ -68,28 +65,117 @@ public static void RemoveOpenTelemetryLibraries(this JsonObject depsJson)
         }
     }
 
+    public static void RemoveOpenTelemetryLibraries(this JsonObject depsJson)
+    {
+        RemoveLibrary(depsJson, lib => lib.StartsWith("OpenTelemetry", StringComparison.Ordinal));
+    }
+
+    public static async Task CleanDuplicatesAsync(this JsonObject depsJson)
+    {
+        var map = DependencyAnalyzer.BuildDependencyMap(depsJson);
+        var framework = depsJson.GetTargetFramework();
+
+        // Analyze matching dependencies between adapter package and supported bytecode instrumentation version range
+        var dependenciesAnalyzeResult = AnalyzeAdapterDependencies(
+            // TODO: Scan these packages from OpenTelemetry.AutoInstrumentation.AdditionalDeps
+            await NugetPackageHelper.GetPackageDependenciesAsync(framework, MongoClientIntegrationMetadata.AdapterAssembly, "1.3.0"),
+            await NugetPackageHelper.GetPackageDependenciesAsync(
+                framework,
+                MongoClientIntegrationMetadata.InstrumentedAssembly,
+                MongoClientIntegrationMetadata.MinimumVersion,
+                MongoClientIntegrationMetadata.MaximumVersion)
+        );
+
+        // Get common packages across versions
+        var commonPackages = GetCommonPackages(dependenciesAnalyzeResult);
+
+        foreach (var package in commonPackages)
+        {
+            // Remove the main library
+            RemoveLibrary(depsJson, lib => lib.Equals(package));
+
+            // Remove transient leftovers
+            DependencyAnalyzer.Cleanup(map, package)
+                .ForEach(transient =>
+                    RemoveLibrary(depsJson, lib => lib.Equals(transient)));
+
+            // TODO: This is just cleaning up deps json, we need to manually
+            // clean shared store also
+        }
+    }
+
+    private static IReadOnlySet<string> GetCommonPackages(Dictionary<NuGetVersion, ICollection<string>> analyzeResult)
+    {
+        var commonPackages = new HashSet<string>(analyzeResult.First().Value);
+
+        foreach (var item in analyzeResult)
+        {
+            commonPackages.IntersectWith(item.Value);
```

Comment:
I'm a bit confuse at this point: it seems that the analysis result has a list for dependencies satisfied by the instrumented assembly for each NuGet framework. However, here it seems that there is an intersection and only packages that were satisfied for every NuGet framework are going to be considered common. It probably will be clear if I see it under the debugger, but, at first glance it is unclear to me.

Reply:
1. Analyser runs per package version (for each package version that exists in the range we support)
2. Makes a final conclusion over the version range.